src/test/fuzz/txrequest.cpp

```diff
@@ -0,0 +1,363 @@
+namespace {
+
+constexpr int MAX_TXHASHES = 16;
+constexpr int MAX_PEERS = 16;
+
+//! Randomly generated GenTxids used in this test (length is MAX_TX).
+uint256 TXHASHES[MAX_TXHASHES];
+
+/** Precomputed random durations (positive and negative, each ~exponentially distributed). */
+std::chrono::microseconds DELAYS[256];
+
+struct Initializer
+{
+    Initializer()
+    {
+        // Use deterministic RNG to fill in txids and delays.
+        // Non-determinism hurts fuzzing.
+        FastRandomContext rng(true);
+        for (int txhash = 0; txhash < MAX_TXHASHES; txhash += 1) {
+            do {
+                TXHASHES[txhash] = rng.rand256();
+            } while (*(TXHASHES[txhash].begin() + 31) != txhash || *(TXHASHES[txhash].begin()) != txhash);
+        }
+        for (int i = 0; i < 16; ++i) {
+            DELAYS[i] = std::chrono::microseconds{i};
+        }
+        for (int i = 16; i < 128; ++i) {
+            DELAYS[i] = DELAYS[i - 1] + std::chrono::microseconds{1 + rng.randbits(((i - 10) * 2) / 9)};
+        }
+        for (int i = 128; i < 256; ++i) {
+            DELAYS[i] = -DELAYS[255 - i];
+        }
+    }
+} g_initializer;
+
+/** Tester class for TxRequestTracker
+ *
+ * It includes a naive reimplementation of its behavior, for a limited set
+ * of MAX_TXHASHES distinct txids, and MAX_PEERS peer identifiers.
+ *
+ * All of the public member functions perform the same operation on
+ * an actual TxRequestTracker and on the state of the reimplementation.
+ * The output of GetRequestable is compared with the expected value
+ * as well.
+ *
+ * Check() calls the TxRequestTracker's sanity check, plus compares the
+ * output of the constant accessors (Size(), CountLoad(), CountTracked())
+ * with expected values.
+ */
+class Tester
+{
+    //! TxRequestTracker object being tested.
+    TxRequestTracker m_tracker;
+
+    //! States for txid/peer combinations in the naive data structure.
+    enum class State {
+        NOTHING, //!< Absence of this txid/peer combination
+
+        // Note that this implementation does not distinguish between BEST/NEW/OTHER variants of CANDIDATE.
+        CANDIDATE,
+        REQUESTED,
+        COMPLETED,
+    };
+
+    //! Sequence numbers, incremented whenever a new CANDIDATE is added.
+    uint64_t m_current_sequence{0};
+
+    //! List of future 'events' (all inserted reqtimes/exptimes). This is used to implement AdvanceToEvent.
+    std::priority_queue<std::chrono::microseconds, std::vector<std::chrono::microseconds>, std::greater<std::chrono::microseconds>> m_events;
+
+    //! Information about a txhash/peer combination.
+    struct Announcement
+    {
+        std::chrono::microseconds m_time;
+        uint64_t m_sequence;
+        State m_state{State::NOTHING};
+        bool m_preferred;
+        bool m_is_wtxid;
+        uint64_t m_priority; //!< Precomputed priority.
+    };
+
+    //! Information about all txhash/peer combination.
+    Announcement m_announcements[MAX_TXHASHES][MAX_PEERS];
```

Comment:
More out of curiosity than a suggestion to change: is there any reason to use C arrays here and elsewhere rather than std::arrays? std::array would allow you to replace many of the hand-written loops below with stl algorithms (e.g. `Cleanup()` becomes a single `std::any()` call and a single `std::transform()` call).

Reply:
All those things work with arrays as well?

Reply:
Ah ok. I guess it's more of a style question really. Is there ever a reason to prefer C arrays over std::array?

Marking this as resolved since it's not important.